src/GafferUI/PlugAdder.cpp

```diff
@@ -0,0 +1,354 @@
+	dragEnterSignal().connect( boost::bind( &PlugAdder::dragEnter, this, ::_2 ) );
+	dragMoveSignal().connect( boost::bind( &PlugAdder::dragMove, this, ::_1, ::_2 ) );
+	dragLeaveSignal().connect( boost::bind( &PlugAdder::dragLeave, this, ::_2 ) );
+	dropSignal().connect( boost::bind( &PlugAdder::drop, this, ::_2 ) );
+	dragEndSignal().connect( boost::bind( &PlugAdder::dragEnd, this, ::_2 ) );
+
+	updateVisibility();
+}
+
+PlugAdder::~PlugAdder()
+{
+}
+
+Imath::Box3f PlugAdder::bound() const
+{
+	return Box3f( V3f( -0.5f, -0.5f, 0.0f ), V3f( 0.5f, 0.5f, 0.0f ) );
+}
+
+void PlugAdder::updateDragEndPoint( const Imath::V3f position, const Imath::V3f &tangent )
+{
+	m_dragPosition = position;
+	m_dragTangent = tangent;
+	m_dragging = true;
+	requestRender();
+}
+
+void PlugAdder::doRender( const Style *style ) const
+{
+	if( m_dragging )
+	{
+		if( !IECoreGL::Selector::currentSelector() )
+		{
+			V3f srcTangent( 0.0f, 0.0f, 0.0f );
+			style->renderConnection( V3f( 0 ), srcTangent, m_dragPosition, m_dragTangent, Style::HighlightedState );
+		}
+	}
+
+	float radius = 0.75f;
+	Style::State state = Style::NormalState;
+	if( getHighlighted() )
+	{
+		radius = 1.25f;
+		state = Style::HighlightedState;
+	}
+	style->renderImage( Box2f( V2f( -radius ), V2f( radius ) ), texture( state ) );
+}
+
+void PlugAdder::addPlug( Gaffer::Plug *connectionEndPoint )
+{
+	UndoContext undoContext( m_node->ancestor<ScriptNode>() );
+
+	if( SwitchComputeNode *switchNode = runTimeCast<SwitchComputeNode>( m_node.get() ) )
+	{
+		switchNode->setup( connectionEndPoint );
+		ArrayPlug *inPlug = switchNode->getChild<ArrayPlug>( "in" );
+		Plug *outPlug = switchNode->getChild<Plug>( "out" );
+
+		StandardNodeGadget::Edge inEdge = StandardNodeGadget::InvalidEdge;
+		if( connectionEndPoint->direction() == Plug::Out )
+		{
+			inPlug->getChild<Plug>( 0 )->setInput( connectionEndPoint );
+			inEdge = m_edge;
+		}
+		else
+		{
+			connectionEndPoint->setInput( switchNode->getChild<Plug>( "out" ) );
+			inEdge = oppositeEdge( m_edge );
+		}
+
+		updateMetadata( inPlug, "nodeGadget:nodulePosition", edgeName( inEdge ) );
+		updateMetadata( inPlug, "compoundNodule:orientation", edgeOrientation( inEdge ) );
+		updateMetadata( outPlug, "nodeGadget:nodulePosition", edgeName( oppositeEdge( inEdge ) ) );
+	}
+}
+
+void PlugAdder::childAdded()
+{
+	updateVisibility();
+}
+
+void PlugAdder::childRemoved()
+{
+	updateVisibility();
+}
+
+void PlugAdder::updateVisibility()
+{
+	if( SwitchComputeNode *switchNode = runTimeCast<SwitchComputeNode>( m_node.get() ) )
+	{
+		setVisible( switchNode->getChild<ArrayPlug>( "in" ) == NULL );
+	}
+}
+
+void PlugAdder::enter( GadgetPtr gadget, const ButtonEvent &event )
+{
+	setHighlighted( true );
+}
+
+void PlugAdder::leave( GadgetPtr gadget, const ButtonEvent &event )
+{
+	setHighlighted( false );
+}
+
+bool PlugAdder::buttonPress( GadgetPtr gadget, const ButtonEvent &event )
+{
+	return event.buttons == ButtonEvent::Left;
```

Comment:
Also, it might be nice if left mouse button press allowed me to setup the switch as a scene/image/filter/color/whatever. Since its a + button its fairly intuitive to want to click it directly.